doc/design/argo_workflow_on_sqlflow.md

```diff
@@ -0,0 +1,66 @@
+# Argo Workflow on SQLFLow
+
+## Overview
+
+SQLFlow translates a SQL program, perhaps with extended SQL syntax for AI, into a workflow. Argo is a Kubernetes native workflow engine. When deploying SQLFlow on Kubernetes, SQLFlow leverages Argo to do workflow management.
+
+When SQLFLow server receives a gRPC `Run` request that contains a SQL program, it:
+
+1. Translates the SQL program to an Argo YAML file.
+1. Submits the YAML file to Kubernetes and receives an Argo workflow ID. `<---(1)`
+1. Returns the workflow ID as a job token to the client.
+
+When SQLFLow server receives a gRPC `Fetch` request contains a job token, it:
+
+1. Looks up the associated Argo workflow and fetches most recent logs. `<---(2)`
```

Comment:
How to define "most recent logs"?

Suppose there are two successive calls to  `Fetch`.  The first call returns logs generated during the period from the starting of the workflow to moment A. How could we implement `Fetch` to make sure that the second call to it returns logs generated after moment A?  Does the SQLFlow server need to keep track of moment A for each workflow ID?  If so, the SQLFlow server keeps workflow status, which is exactly what we DON'T want and is the reason we introduce Argo, an execution engine to run workflows.

Reply:
[Kubernetes Logs API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#read-log) supports the `sinceTime` parameter to fetch logs from the specified timestamp,  the client side keeps`FetchToken` object which uses this `timestamp` as `logOffset` to fetch logs in an incremental way, https://github.com/sql-machine-learning/sqlflow/issues/1382 is the more detailed survey.